File: src/main/java/xyz/elmot/interpret/eval/ProgVisitor.java
package xyz.elmot.interpret.eval;

import xyz.elmot.interpret.AtorBaseVisitor;
import xyz.elmot.interpret.AtorParser;

import java.util.LinkedHashMap;
import java.util.Map;
import java.util.function.Consumer;

public class ProgVisitor extends AtorBaseVisitor<Void> {
    public ProgVisitor(Consumer<String> out) {
        this.out = out;
    }

    private final Consumer<String> out;
    private Map<String, Value> vars = new LinkedHashMap<>();

    @Override
    public Void visitPrint(AtorParser.PrintContext ctx) {
        String text = ctx.TEXT().getText();
        out.accept(text.substring(1, text.length() - 1));
        return null;
    }

    @Override
    public Void visitStmt(AtorParser.StmtContext ctx) {
        checkCancel();
        return super.visitStmt(ctx);
    }

    @SuppressWarnings("WeakerAccess")
    public static void checkCancel() {
        if (Thread.currentThread().isInterrupted()) {
            throw new CancelException();
        }
    }

    @Override
    public Void visitVar(AtorParser.VarContext ctx) {
        String name = ctx.NAME().getText();
        ExprVisitor exprVisitor = new ExprVisitor(vars);
        AtorParser.ExprContext expr = ctx.expr();
        expr.accept(exprVisitor);
        vars.put(name, exprVisitor.getResult(expr));
        return null;
    }

    @Override
    public Void visitOut(AtorParser.OutContext ctx) {
        ExprVisitor exprVisitor = new ExprVisitor(vars);
        AtorParser.ExprContext expr = ctx.expr();
        expr.accept(exprVisitor);
        out.accept(exprVisitor.getResult(expr).getString());
        return null;
    }
}

File: src/main/kotlin/xyz/elmot/interpret/eval/ProgVisitor.kt
package xyz.elmot.interpret.eval

import xyz.elmot.interpret.AtorBaseVisitor
import xyz.elmot.interpret.AtorParser
import java.util.*

class ProgVisitor(val out: (String)->Unit) : AtorBaseVisitor<Any>() {
    /**
     * Global script variables
     */
    private val vars = LinkedHashMap<String, Value>()

    override fun visitPrint(ctx: AtorParser.PrintContext): Any? {
        val text = ctx.TEXT().text
        out.invoke(text.substring(1, text.length - 1))
        return null
    }

    override fun visitStmt(ctx: AtorParser.StmtContext): Any? {
        checkCancel()
        return super.visitStmt(ctx)
    }

    override fun visitVar(ctx: AtorParser.VarContext ):Any? {
        val name = ctx.NAME().text
        val value = ExprVisitor.calcValue(ctx.expr(), vars)
        vars.put(name, value)
        return null
    }

    override fun visitOut(ctx: AtorParser.OutContext ):Any? {
        val value = ExprVisitor.calcValue(ctx.expr(), vars)
        out.invoke(value.getString())
        return null
    }

    companion object {
        fun checkCancel() {
            if (Thread.interrupted()) {
                throw CancelException()
            }
        }

    }
}